pkg/didmethod/peer/did.go

```diff
@@ -30,7 +32,59 @@ func newDid(doc *did.Doc) (string, error) {
 	if doc.PublicKey == nil || doc.Authentication == nil {
 		return "", errors.New("the genesis version must include public keys and authentication")
 	}
+	numBasis, err := hashString(doc)
+	if err != nil {
+		return "", err
+	}
+	messageIdentifier := []string{peerPrefix, numAlgo, encAlgo, "-", numBasis}
+	peerDID := strings.Join(messageIdentifier, "")
+
+	return peerDID, nil
+}
+
+func validate(doc *did.Doc) error {
+	did := doc.ID
+	hashedNumBasis, err := validateDid(did)
+	if err != nil {
+		return err
+	}
+	//removing did to calculate the hash of genesis doc
+	doc.ID = ""
+	numBasis, err := hashString(doc)
+	if err != nil {
+		return err
+	}
+
+	if !(numBasis == hashedNumBasis) {
+		return errors.New("hash of the doc doesnt match the computed hash")
+	}
+	return nil
+}
+
+func validateDid(did string) (string, error) {
+	//no whitespace
+	for _, v := range did {
+		// test each character to see if it is whitespace.
+		if unicode.IsSpace(v) {
+			if v == 32 {
+				return "", errors.New("white space is not allowed in did")
+			}
+		}
+	}
+	// matching regex
+	matched, _ := regexp.MatchString(`did:peer:(1)(1)-([a-fA-F0-9]{64})*`, did)
+	if !matched {
+		return "", errors.New("did doesnt follow matching regex")
+	}
+
+	hashed := strings.FieldsFunc(did, split)
+
+	numBasis := hashed[1]
 
+	return numBasis, nil
+}
+
+func hashString(doc *did.Doc) (string, error) {
```

Comment:
This can be renamed to "numBasis" to mirror the terms used in the spec.

Reply:
refactored